scipy/stats/_hypotests.py

```diff
@@ -663,3 +666,361 @@ def somersd(x, y=None):
+        inoculated with a placebo. Twelve of the 15 subjects in the placebo
+        group (80%) eventually became infected with influenza whereas for the
+        vaccine group, only 7 of the 15 subjects (47%) became infected. The
+        data are tabulated as a 2 x 2 table::
+
+                Vaccine  Placebo
+            Yes     7        12
+            No      8        3
+
+    When working with statistical hypothesis testing, we usually use a
+    threshold probability or a significance level upon which we decide
+    to reject or not the null hypothesis :math:`H_0`. A commonly used
+    significance level is 5%. In this example, we are using the
+    `alternative` parameter with the "less" option, because the vaccine has
+    either no effect (:math:`H_0` hypothesis) or a positive effect
+    (:math:`H_1` hypothesis). Therefore, under the null hypothesis
+    (:math:`H_0`), the probability :math:`p_2` of catching the virus
+    **without** any vaccine, is either equal or lower than :math:`p_1`,
+    the probability of catching the virus **with** the vaccine.
+    To compute the p-value with Banard test, let's call `barnard_exact` as
+    follow :
+
+    >>> import scipy.stats as stats
+    >>> res = stats.barnard_exact([[7, 12], [8, 3]], alternative="less")
+    >>> res.statistic
+    -1.894...
+    >>> res.pvalue
+    0.03407...
+
+    Then, the probability of obtaining test results at least as extreme as the
+    data observed above is around 3.4%. Since our p-value is under our
+    significance level defined above, we reject :math:`H_0`. The vaccine has
+    a positive effect and we have :math:`p_1 \leq p_2`; the probability of
+    contracting the virus if we have been vaccinated is lower than if
+    we are not.
+
+    We can compare with Fisher's exact test which produces an exact
+    p-value of 6.4% :
+
+    >>> _, pvalue = stats.fisher_exact([[7, 12], [8, 3]], alternative="less")
+    >>> pvalue
+    0.0640...
+
+    With the same threshold probability of 5%, `fisher_exact` is accepting
+    :math:`H_0` while `barnard_exact` is rejecting it. As stated in [2]_,
+    Barnard's test is uniformly more powerful than Fisher's exact test
+    because Barnard's test do not condition on any margin. Fisher's test
+    should only be used when both sets of marginals are fixed.
+
+    """
+    if n <= 0:
+        raise ValueError(
+            "Number of points `n` must be strictly positive, "
+            f"found {n!r}"
+        )
+
+    table = np.asarray(table, dtype=np.int64)
+
+    if not table.shape == (2, 2):
+        raise ValueError("The input `table` must be of shape (2, 2).")
+
+    if np.any(table < 0):
+        raise ValueError("All values in `table` must be nonnegative.")
+
+    if 0 in table.sum(axis=0):
+        # If both values in column are zero, the p-value is 1 and
+        # the score's statistic is NaN.
+        return BarnardExactResult(np.nan, 1.0)
```

Comment:
Consider adding a test for this. Is a zero row an edge case or is that expected to be
```
>>> stats.barnard_exact([[0, 0],[3, 4]])
BarnardExactResult(statistic=0.0, pvalue=1.0)
```
? I ask because
```
>>> stats.fisher_exact([[0, 0],[3, 4]])
(nan, 1.0)
```
In any case, consider adding a test for that, too, just so it's clear that we checked.

Reply:
In `fisher_exact` we have :
```Python
    if 0 in c.sum(axis=0) or 0 in c.sum(axis=1):
        # If both values in a row or column are zero, the p-value is 1 and
        # the odds ratio is NaN.
        return np.nan, 1.0
```
because the statistic (which is in this case the oddration) is defined as follow:
```Python
oddsratio = c[0, 0] * c[1, 1] / (c[1, 0] * c[0, 1])
```
It is thus clear that if we have one row or one column which sums equal to zero, then we have 0 / 0 which leads to NaN. 
But in our case, the only edge case I see is when one of the column is equal to zero. So I guess 
```Python
>>> stats.barnard_exact([[0, 0],[3, 4]])
BarnardExactResult(statistic=0.0, pvalue=1.0)
```
is the expected result. I am going to add a test nevertheless 👍

Reply:
For the math pov, the Fisher exact contingency test is not defined if a row sums to zero or a column sum to zero, because it explore hypergeometric dists conditionnaly to rows sums and cols sums.

The Barnard test explore pairs of binomials conditionnaly to cols sums. Therefore, the Barnard test is perfectly definied when a row sums is zero. However, in this case, the stat is 0 and the pvalue is 1. The test is defined, but it is completly useless.

Reply:
Great.